py/sentry_streams/pipeline.py

```diff
@@ -114,5 +126,21 @@ class Map(WithInput):
     # instead of a raw string
     # TODO: Allow product to both enable and access
     # configuration (e.g. a DB that is used as part of Map)
-    function: str
+    function: Callable[..., Any]
     step_type: StepType = StepType.MAP
+
+
+@dataclass
+class Reduce(WithInput):
+    # group_by_key: refactor to Callable reference
+    group_by_key: Callable[..., Any]
+    # windowing mechanism, is this going to be mandatory?
+    # windowing: Window
+    # aggregation (use standard accumulator)
+    aggregate_fn: Accumulator
+    step_type: StepType = StepType.REDUCE
+    # storage: a fixed (enum?) set of storage backends we provide
+    # consider making this a class
+    storage: StateBackend = StateBackend.HASH_MAP
```

Comment:
Isn't the storage hidden by the accumulator at times ?
Think about these cases:
- Aggregation with windowing and producing the output. Here we can separate aggregaiton function from store as we can store the aggregation state where we want.
- Global state: accumulating counts in postgres. Here the storaeg is application specific so there is no point in separating the two.

Reply:
I _think_ this is the same discussion we had related to https://github.com/getsentry/streams/pull/18#discussion_r1964903269

I can leave a TODO to come back to this in a next PR since this once is getting large

Reply:
> Global state: accumulating counts in postgres. Here the storaeg is application specific so there is no point in separating the two.

We could separate the accumulation logic from the flush to postgres concerns. We may also have scenarios where we need to flush streaming aggregations to redis.

Reply:
I've taken a stab at this via the `AggregationBackend`. Right now, the backend can take the output of the Accumulator and flush it to a configurable storage. 

Let me know what you all think of this interface.